crates/ra_analysis/src/goto_defenition.rs

```diff
@@ -41,6 +42,22 @@ pub(crate) fn reference_defenition(
             return Ok(vec![nav]);
         };
     }
+    // Then try module name resolution
+    if let Some(module) =
+        hir::source_binder::module_from_child_node(db, file_id, name_ref.syntax())?
+    {
+        if let Some(path) = name_ref
+            .syntax()
+            .ancestors()
+            .find_map(ast::Path::cast)
+            .and_then(hir::Path::from_ast)
+        {
+            let resolved = module.resolve_path(db, &path)?;
+            if let Some(def_id) = resolved.take_types().or(resolved.take_values()) {
+                return Ok(vec![navigation_target_for_def(db, def_id)]);
+            }
+        }
+    }
```

Comment:
It also seems that "if local, resolve local, else resolve in module" should be in `hir`... Perhaps function scopes need to be taught to resolve stuff to either `Binding` or `DefId`?

Reply:
Hm, it's a bit complicated since we want to be able to do the module scope resolution also outside of functions. Also, designing the API for this correctly seems non-trivial because for example if we have `some::pa<|>th::to::somewhere`, I think goto def should go to the module `some::path`, not to the resolution of the full path. (Maybe that's wrong though.)

Reply:
Hm, makes sense! Basically the correct logic here is precisely “thing should be resolved in the module scope if it is not resolved at a local scope”. I think this might change though once we support function-local items, but maybe not?